libopae/plugins/xfpga/opae_drv.c

```diff
@@ -443,68 +481,396 @@ fpga_result dfl_fme_port_reset(int fd)
+			if (!strncmp(OPAE_FME, dirent->d_name + type_beg,
+				     OPAE_FME_LEN)) {
+				device->fme = make_opae_resource(
+					device, dirent->d_name, num,
+					FPGA_DEVICE);
+			} else if (!strncmp(OPAE_PORT,
+					    dirent->d_name + type_beg,
+					    OPAE_PORT_LEN)) {
+				device->port = make_opae_resource(
+					device, dirent->d_name, num,
+					FPGA_ACCELERATOR);
+			}
+		}
+	}
+	regfree(&re);
+	closedir(dir);
+}
+
+STATIC void discover()
+{
+	int res = 0;
+	uint32_t i = 0;
+	char err[128];
+	DIR *dir = NULL;
+	struct dirent *dirent = NULL;
+	const char *sysfs_class_path = _drv->sysfs.class_path;
+	regex_t device_re, pcie_re;
+	regmatch_t matches[RE_REGION_GROUPS];
+	int reg_res = regcomp(&device_re, _drv->sysfs.region_fmt,
+			      REG_EXTENDED);
+	if (reg_res) {
+		regerror(reg_res, &device_re, err, 128);
+		FPGA_ERR("Error compling regex: %s", err);
+		return;
+	};
+
+	reg_res = regcomp(&pcie_re, PCIE_PATH_PATTERN, REG_EXTENDED | REG_ICASE);
+	if (reg_res) {
+		regfree(&device_re);
+		FPGA_ERR("Error compling regex");
+		return;
+	}
+
+	dir = opendir(sysfs_class_path);
+	while ((dirent = readdir(dir))) {
+		if (!strcmp(dirent->d_name, "."))
+			continue;
+		if (!strcmp(dirent->d_name, ".."))
+			continue;
+		// if the current directory matches the device regex
+		reg_res = regexec(&device_re, dirent->d_name, RE_REGION_GROUPS,
+				  matches, 0);
+		if (!reg_res) {
+			int num_begin = matches[1].rm_so;
+			if (num_begin < 0) {
+				FPGA_ERR("sysfs format invalid: %s", dirent->d_name);
+				continue;
+			}
+			int num = strtoul(dirent->d_name + num_begin, NULL, 10);
+			// increment our device count after filling out details
+			// of the discovered region in our _devices array
+			if (opae_mutex_lock(res, &_sysfs_device_lock)) {
+				goto out_free;
+			}
+			if (add_device(&_devices[_sysfs_device_count++],
+					sysfs_class_path, dirent->d_name,
+					num, &pcie_re)) {
+				FPGA_MSG("Error processing device: %s",
+					 dirent->d_name);
+				_sysfs_device_count--;
+			}
+			if (opae_mutex_unlock(res, &_sysfs_device_lock)) {
+				goto out_free;
+			}
+		}
+	}
+
+	if (opae_mutex_lock(res, &_sysfs_device_lock)) {
+		goto out_free;
+	} else if (!_sysfs_device_count) {
+		FPGA_ERR("Error discovering fpga regions");
+		goto out_unlock;
+	}
+
+	// now, for each discovered region, look inside for resources
+	// (fme|port)
+	for (i = 0; i < _sysfs_device_count; ++i) {
+		sysfs_find_resources(&_devices[i]);
+	}
+
+out_unlock:
+	if (pthread_mutex_unlock(&_sysfs_device_lock)) {
+		FPGA_MSG("error unlocking sysfs region mutex");
+	}
+out_free:
+	regfree(&device_re);
+	regfree(&pcie_re);
+	closedir(dir);
+}
+
+STATIC int sysfs_device_destroy(opae_device *device)
+{
+	ASSERT_NOT_NULL(device);
+	if (device->fme) {
+		free(device->fme);
+		device->fme = NULL;
+	}
+	if (device->port) {
+		free(device->port);
+		device->port = NULL;
+	}
+	return FPGA_OK;
+}
+
+int opae_drv_initialize(void)
+{
+	opae_drv_finalize();
```

Comment:
I intend to leave it as it fixes stuff when things finalize isn't called. There are a few tests that don't call finalize. Should I fix the tests instead or leave this as a fail-safe?

Reply:
eh, it just struck me as a bit strange.  I think it would read better if it was removed, but if it serves a purpose.  Probably at least worth a comment.

Reply:
👍